### ISSUE TITLE
Directory rename deletes children whose row keys are unchanged

### ISSUE THREAD

**Issue by milandesai** [GiraffaFS/giraffa#230]
When a directory is renamed, we check the row key of the source and destination paths. If the row key stayed the same, we skip the "copy then delete" procedure. This is because with the introduction of numbered row keys in #34, the row key is no longer unique per path. If we didn't do the check and the row key didn't change, the "copy" phase of the procedure would not actually create a namespace table row, and then the "delete" phase would delete the existing row, effectively deleting the file/directory instead of renaming it.

While we do this check for the root paths of the source and destination directories, we don't do this check for the children. It is possible that the row key changes for the root path but not for some of the children. For example, with id-based row keys with a depth of 2, consider the rename of /a/b/c to /a/d/c. The source row key is id(b)+id(c) and the destination row key is id(d)+id(c). Since the row keys are different, we will proceed to the "copy then delete" recoverable rename procedure. Now consider a child of directory c. /a/b/c/f is renamed to /a/d/c/f; both the source and destination row keys are id(c)+id(f). We will first copy /a/b/c/f to /a/d/c/f, but both paths will point to the same namespace table entry. Later we will delete /a/b/c/f, and by doing so we will also unintentionally delete destination file /a/d/c/f since it uses the same row key. In fact, by moving /a/b/c to /a/d/c, every single child of directory c will be deleted.

I can think of two possible solutions:
1. Don't delete any nodes that have the "renaming" flag set. This solution is pretty simple but may be unintuitive.
2. Don't bulk copy / bulk delete; for each child node, copy each INode then delete it before moving on to the next INode. Check for unchanged row-keys before doing the copy/delete. This solution is more involved and less efficient, but more intuitive.

**Comment by milandesai** [comment]
Attached a pull request that implements solution (1) above.

**Comment by pjeli** [comment]
Couple of things to clarify about solutions:

For (1), we are safe if using Full Path because a/b/c/f and a/d/c/f will lead to different RowKeys, right? Meaning we don't risk deleting the thing we are supposed to be copying. I am confirming we haven't broken support for FullPathRowKey here. Perhaps we want some way to clarify that (1) is specifically for NumberedRowKey.

For (2), what do you mean by "don't bulk"? From looking over how it is now I don't believe we do any bulking apart from maybe final directory deletion(s). It looks like files are already being copied and then deleted as (2) proposes.

**Comment by milandesai** [comment]
Yes, this bug would only occur with numbered row keys since full path row keys are unique per path. What (2) is proposing is that for each node that needs to be renamed, we copy it, then delete it, then move on to the next node. What we currently do is copy all nodes, then delete them. With the former approach, we can check each node and see if we need to do anything. With the current approach, we would have to either mark such nodes as not needing to be deleted (i.e. via the rename state flag), or make a list of the ones that do need to be deleted rather than deleting the whole parent directory.

You're right that we don't bulk copy, we do a partial bulk delete though - we bulk delete the directory nodes but not the file nodes.

**Comment by pjeli** [comment]
Thanks Milan I think I understand now. You are right; we do a phase of copy and then a phase of delete, not copy-delete copy-delete etc, that's what I was not piecing together.

I am for (1) but it would be nice to have some way of saying that it is specific to Numbered RowKey. I am weary that someone may want to implement XYZ RowKey, which (1) may break if we aren't careful.

It may be that (1) is safe under all circumstances.. I need some time to think it through.
